app/abci/utils.go

```diff
@@ -0,0 +1,232 @@
+	commitInfo := ctx.CometInfo().GetLastCommit()
+
+	// Check that both extCommit + commit are ordered in accordance with vp/address.
+	if err := ValidateExtendedCommitAgainstLastCommit(extCommit, commitInfo); err != nil {
+		return err
+	}
+
+	// Start checking vote extensions only **after** the vote extensions enable
+	// height, because when `currentHeight == VoteExtensionsEnableHeight`
+	// PrepareProposal doesn't get any vote extensions in its request.
+	extensionsEnabled := VoteExtensionsEnabled(ctx)
+	marshalDelimitedFn := func(msg proto.Message) ([]byte, error) {
+		var buf bytes.Buffer
+		if err := protoio.NewDelimitedWriter(&buf).WriteMsg(msg); err != nil {
+			return nil, err
+		}
+
+		return buf.Bytes(), nil
+	}
+
+	var (
+		// Total voting power of all vote extensions.
+		totalVP int64
+		// Total voting power of all validators that submitted valid vote extensions.
+		sumVP int64
+	)
+
+	for _, vote := range extCommit.Votes {
+		totalVP += vote.Validator.Power
+
+		if extensionsEnabled {
+			if vote.BlockIdFlag == cmtproto.BlockIDFlagCommit && len(vote.ExtensionSignature) == 0 {
+				return fmt.Errorf("vote extension signature is missing; validator addr %s",
+					vote.Validator.String(),
+				)
+			}
+			if vote.BlockIdFlag != cmtproto.BlockIDFlagCommit && len(vote.VoteExtension) != 0 {
+				return fmt.Errorf("non-commit vote extension present; validator addr %s",
+					vote.Validator.String(),
+				)
+			}
+			if vote.BlockIdFlag != cmtproto.BlockIDFlagCommit && len(vote.ExtensionSignature) != 0 {
+				return fmt.Errorf("non-commit vote extension signature present; validator addr %s",
+					vote.Validator.String(),
+				)
+			}
+		} else { // vote extensions disabled
+			if len(vote.VoteExtension) != 0 {
+				return fmt.Errorf("vote extension present but extensions disabled; validator addr %s",
+					vote.Validator.String(),
+				)
+			}
+			if len(vote.ExtensionSignature) != 0 {
+				return fmt.Errorf("vote extension signature present but extensions disabled; validator addr %s",
+					vote.Validator.String(),
+				)
+			}
+
+			continue
+		}
+
+		// Only check + include power if the vote is a commit vote. There must be super-majority, otherwise the
+		// previous block (the block vote is for) could not have been committed.
+		if vote.BlockIdFlag != cmtproto.BlockIDFlagCommit {
+			continue
+		}
+
+		// If the validator does not have a valid public key, we skip the signature verification logic but still include
+		// the validator's voting power in the total voting power. The app may have pruned the validator's public key
+		// from the store, but comet considered the validator as active and included them in the commit since there
+		// is a 1 block delay between the validator set update on the app and comet.
+		sumVP += vote.Validator.Power
+		valConsAddr := sdk.ConsAddress(vote.Validator.Address)
+		pubKeyProto, err := valStore.GetPubKeyByConsAddr(ctx, valConsAddr)
+		if err != nil {
+			continue
+		}
```

Comment:
I'm unsure why 
(1) we add `sumVP` here when the signature verification hasn't happened
(2) the default SDK doesn't account for the scenario described in the comment.

Reply:
I'm also not sure why, I think we can stick more closely to the default implementation.